paper/paper.Rmd

```diff
@@ -0,0 +1,221 @@
+      variant: "markdown"
+      preserve_yaml: true
+      standalone: true
+bibliography: paper.bib
+csl: apa.csl
+link-citations: yes
+---
+
+```{r setup, warning=FALSE, message=FALSE, echo=FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  out.width = "100%",
+  comment = "#>"
+)
+
+library(lintr)
+
+withr::local_options(list(
+  lintr.format_width = 60L
+))
+```
+
+# Statement of Need
+
+R is an interpreted, dynamically-typed programming language [@base2023]. It is a popular choice for statistical analysis and visualization, and is used by a wide range of researchers and data scientists. The `{lintr}` package is an open-source R package that provides static code analysis to check for a variety of common problems related to readability, efficiency, consistency, style, etc. In particular, by default it enforces the tidyverse style guide [@Wickham2023]. It is designed to be easy to use and integrate into existing workflows, and can be used as part of an automated build or continuous integration process. `{lintr}` also integrates with a number of popular IDEs and text editors, such as RStudio and Visual Studio Code, making it convenient for users to run `{lintr}` checks on their code as they work.
+
+# Features
+
+As of this writing, `{lintr}` offers `r length(all_linters())` linters.
+
+```{r all_linters}
+library(lintr)
+
+length(all_linters())
+```
+
+Naturally, we can't discuss all of them here. To see details about all available linters, we encourage readers to see <https://lintr.r-lib.org/dev/reference/index.html#individual-linters>.
+
+We will showcase one linter for each kind of common problem found in R code.
+
+- **Best practices**
+
+`{lintr}` offers linters that can detect problematic antipatterns and suggest alternatives that follow best practices.
+
+For example, expressions like `ifelse(x, TRUE, FALSE)` and `ifelse(x, FALSE, TRUE)` are redundant; just `x` or `!x` suffice in R code where logical vectors are a core data structure. The `redundant_ifelse_linter()` linter detects such discouraged usages.
+
+```{r redundant_ifelse_linter}
+lint(
+  text = "ifelse(x >= 2.5, TRUE, FALSE)",
+  linters = redundant_ifelse_linter()
+)
+```
+
+- **Efficiency**
+
+Sometimes the users might not be aware of a more efficient way offered by R for carrying out a computation. `{lintr}` offers linters to improve code efficiency by avoiding common inefficient patterns.
+
+For example, the `any_is_na_linter()` linter detects usages of `any(is.na(x))` and suggests `anyNA(x)` as a more efficient alternative to detect presence of *any* missing values.
+
+```{r any_is_na_linter_with_lint}
+lint(
+  text = "any(is.na(x), na.rm = TRUE)",
+  linters = any_is_na_linter()
+)
+```
+
+`anyNA()` in R is more efficient than `any(is.na())` because it stops execution once a missing value is found, while `is.na()` evaluates the entire vector.
+
+```{r any_is_na_linter_without_lint}
+lint(
+  text = "anyNA(x)",
+  linters = any_is_na_linter()
+)
+```
```

Comment:
Below, we don't offer the "bad"/"good" pairing of snippets, should we remove this "good" example here?

Reply:
But this chunk is a companion to `any_is_na_linter_with_lint` chunk, which comes right before it.

Reply:
Right... there's still some inconsistency. We have companion chunks for `comparison_negation_linter()` and `any_is_na_linter()`, but not `object_name_linter()` or `redundant_ifelse_linter()`. Since the corresponding "good" code for `redundant_ifelse_linter()` is covered in the text, maybe it's only `object_name_linter()` that's the outlier? Or do you think it's fine as is?

Reply:
Good point. I hadn't included the no-lint versions for these because they felt too trivial, but it does feel inconsistent to not see them. I have included them now.